### PR TITLE
Zone-based pass generator Python bindings and notebook visualizations

### LINKED ISSUE #2056: Python bindings and Jupyter visualization for zone-based pass generator
### Description of the task

Python bindings and Jupyter visualization for zone-based pass generator

### Acceptance criteria

- [ ] python bindings
- [ ] cool visualization

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

image

- Plot the pass generator zones
- Plot the best pass for each zone (mouseover the receiver point to see pass rating)
- Line segment between passer and receiver points becomes more visible as the pass quality increases

### Testing Done

See picture

### Resolved Issues

resolves #2056 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/python_bindings/pass_generator.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+py::list getBestPassesForAllZones(EighteenZonePassGenerator& generator,
+                                  const World& world)
+{
+    py::list result;
+    auto evaluation = generator.generatePassEvaluation(world);
+
+    // unfortunately this is tightly coupled to the 18 zone field division
+    for (auto zone_id = static_cast<int>(EighteenZoneId::ZONE_1);
+         zone_id <= static_cast<int>(EighteenZoneId::ZONE_18); zone_id++)
+    {
```
Comment (akhilveeraghanta) [inline src/software/python_bindings/pass_generator.cpp]: Could you just do
`for (ZoneEnum zone_id : pitch_division->getAllZoneIds())`?

File: src/software/python_bindings/pass_generator.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+py::list getAllZones(const World& world)
+{
+    EighteenZonePitchDivision pitch_division(world.field());
+    py::list result;
+    for (auto zone_id = static_cast<int>(EighteenZoneId::ZONE_1);
+         zone_id <= static_cast<int>(EighteenZoneId::ZONE_18); zone_id++)
+    {
+        result.append(pitch_division.getZone(static_cast<EighteenZoneId>(zone_id)));
+    }
+    return result;
```
Comment (akhilveeraghanta) [inline src/software/python_bindings/pass_generator.cpp]: You can probably replace this with `for (ZoneEnum zone_id : pitch_division->getAllZoneIds())`?

File: src/software/python_bindings/pass_utilities.h [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+#include <pybind11/pybind11.h>
+
+#include "software/ai/passing/pass.h"
+
+Pass createPassFromDict(pybind11::dict pass_dict);
+
+pybind11::dict convertPassToDict(const Pass& pass);
```
Comment (akhilveeraghanta) [inline src/software/python_bindings/pass_utilities.h]: nit: would be nice to have some java docs here even though its pretty obvious what these functions are doing

File: src/software/python_bindings/pass_generator.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+/**
+ * Gets the best pass for each of the 18 zones and returns them in a list of dicts
+ * as per pass_utilities.cpp.
+ * @param generator an EighteenZonePassGenerator
+ * @param world the world
+ * @return a list of dicts describing the best pass in each of the zones.
+ */
+py::list getBestPassesForAllZones(EighteenZonePassGenerator& generator,
```
Comment (akhilveeraghanta) [inline src/software/python_bindings/pass_generator.cpp]: this can probably take any pass generator, see comment below

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
overall looks pretty good! just a couple comments

**Review by kelvinkoon — COMMENTED** [review]
LGTM, nothing to add besides Akhil's comments.

File: src/python_tools/plotting/plot_pass_generator.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            passes_dict["receiver_xs"].append(receiver_point.x())
+            passes_dict["receiver_ys"].append(receiver_point.y())
+
+            passes_dict["pass_line_xs"].append([passer_point.x(), receiver_point.x()])
+            passes_dict["pass_line_ys"].append([passer_point.y(), receiver_point.y()])
+
+            passes_dict["pass_rating"].append(rating)
+            # adjusted pass rating is only used to determine the alpha of the
+            # line segment between passer and receiver points
+            passes_dict["adjusted_pass_rating"].append(np.clip(rating * 2, 0, 1))
```
Comment (jonathanlew) [inline src/python_tools/plotting/plot_pass_generator.py]: Changing the alpha makes bad passes hard to see, right? Could we change the brightness instead, i.e. dark orange to light orange.

**Review by jonathanlew — COMMENTED** [review]
Not much to add, very cool!

**Comment by jordanzliu** [comment]
image
Made some minor visualization changes, here's a preview

File: src/software/python_bindings/pass_generator.cpp (hunk 4 above)
Comment (jordanzliu) [reply src/software/python_bindings/pass_generator.cpp]: function needs to be template of `ZoneEnum` type and so it still can't accept any `PassGenerator`, but made the change anyways.

File: src/python_tools/plotting/plot_pass_generator.py (hunk 5 above)
Comment (jordanzliu) [reply src/python_tools/plotting/plot_pass_generator.py]: made some visualization changes as below

**Review by jonathanlew — COMMENTED** [review]
> image
> Made some minor visualization changes, here's a preview



Nice visualization! Just one little comment, can the colour bar on the right be labelled, i.e. pass score

File: src/software/python_bindings/pass_generator.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+ * @param world the world
+ * @return a list of dicts describing the best pass in each of the zones.
+ */
+template <typename ZoneEnum>
+py::list getBestPassesForAllZones(PassGenerator<ZoneEnum>& generator, const World& world)
+{
+    py::list result;
+    auto evaluation = generator.generatePassEvaluation(world);
+
+    // unfortunately this is tightly coupled to the 18 zone field division
```
Comment (akhilveeraghanta) [inline src/software/python_bindings/pass_generator.cpp]: Is this comment still true?

**Review by akhilveeraghanta — COMMENTED** [review]
lgtm! just one more comment

File: src/software/python_bindings/pass_generator.cpp (hunk 6 above)
Comment (jordanzliu) [reply src/software/python_bindings/pass_generator.cpp]: nope, removed comment

**Comment by jordanzliu** [comment]
> > image
> > Made some minor visualization changes, here's a preview
> 
> Nice visualization! Just one little comment, can the colour bar on the right be labelled, i.e. pass score

added a title to it, but it messes up the layout a little bit

**Review by akhilveeraghanta — APPROVED** [review]
💯

**Review by jonathanlew — APPROVED** [review]
:shipit:

**Pull request merged**